### ISSUE TITLE
an npm package is able to break core functionality (Linking.getInitialURL promise never resolves)

### ISSUE THREAD

**Issue by aforty** [facebook/react-native#31558]
## Description
There is a npm package (`json-api-normalizer`) that is able to break core functionality. Just by importing it the `Linking.getInitialURL()` promise never resolves. I suspect the change in this commit [this commit](https://github.com/yury-dymov/json-api-normalizer/commit/8158e433bb3d8f2498a96511260a4a21b60e5648) breaks this because it works with 1.0.1 but not >1.0.2. I'd like to know exactly what the problem is so that I can work with the package maintainer to fix this.

## React Native version:
```
info Fetching system and libraries information...
System:
    OS: macOS 11.4
    CPU: (6) x64 Intel(R) Core(TM) i5-9600K CPU @ 3.70GHz
    Memory: 434.93 MB / 64.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.15.4 - ~/.nvm/versions/node/v14.15.4/bin/node
    Yarn: 1.22.4 - ~/.nvm/versions/node/v14.15.4/bin/yarn
    npm: 6.14.10 - ~/.nvm/versions/node/v14.15.4/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.10.1 - /Users/aforty/.rvm/gems/ruby-2.7.1/bin/pod
  SDKs:
    iOS SDK:
      Platforms: iOS 14.5, DriverKit 20.4, macOS 11.3, tvOS 14.5, watchOS 7.4
    Android SDK:
      API Levels: 23, 24, 25, 26, 27, 28, 29
      Build Tools: 27.0.3, 28.0.3, 29.0.2, 29.0.3, 30.0.0, 30.0.3
      System Images: android-30 | Google APIs Intel x86 Atom, android-30 | Google Play Intel x86 Atom
      Android NDK: 22.1.7171670
  IDEs:
    Android Studio: 4.2 AI-202.7660.26.42.7322048
    Xcode: 12.5/12E262 - /usr/bin/xcodebuild
  Languages:
    Java: 1.8.0_144 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.1 => 17.0.1
    react-native: 0.64.1 => 0.64.1
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

## Steps To Reproduce
Provide a detailed list of steps that reproduce the issue.

1. Initialize: `npx react-native init JsonApiNormalizerBug && cd JsonApiNormalizerBug && yarn add json-api-normalizer`
2. Replace App.js with the code below.
3. Run: `npx react-native run-ios`
4. Notice `didFinish` never turned true and thus it says `Linking.getInitialURL() is loading.` 
5. Notice if you comment out the first line (import `json-api-normalizer`) then the promise resolves correctly.

```javascript
// App.js

import normalize from 'json-api-normalizer';
import React, {useEffect, useState} from 'react';
import {Linking, SafeAreaView, Text, View} from 'react-native';

export default function App() {
  const [didFinish, setDidFinish] = useState(false);

  useEffect(() => {
    Linking.getInitialURL()
      .then(() => setDidFinish(true))
      .catch(() => setDidFinish(true));
  }, []);

  return (
    <SafeAreaView>
      <View>
        {didFinish ? (
          <Text>Linking.getInitialURL() finished.</Text>
        ) : (
          <Text>Linking.getInitialURL() is still loading.</Text>
        )}
      </View>
    </SafeAreaView>
  );
}
```

## Expected Results
A compiled npm package should not be able to break react-native in this way. 

If it's not a problem in react-native or unable to fix, what is should the package maintainer do to remedy this?

**Comment by jhclaura** [comment]
Also have this issue after upgrading React Native from 0.63.4 to 0.64.1, not knowing any specific NPM package, just all Promises never resolve, would it be something missing during the upgrade?

**Comment by petercarpenter-santander** [comment]
I experienced a similar issue after upgrading from 61.5 to 64.2, no promises in my app would resolve. I tracked the issue down to the change in `metro.config.js` ([new file](https://raw.githubusercontent.com/react-native-community/rn-diff-purge/release/0.64.2/RnDiffApp/metro.config.js)) - if I change `inlineRequires` back to false, the issue with the promises goes away.

I assume there is some conflict between one of my dependencies and the way `inlineRequires` works. As its only a performance optimisation I'm just going to leave that turned off for now, but would be good to understand why that's happening.
